cmd/juju/cloud/show.go

```diff
@@ -103,56 +110,131 @@ func (c *showCloudCommand) Init(args []string) error {
 	default:
 		return errors.New("no cloud specified")
 	}
-	var err error
-	c.ControllerName, err = c.ControllerNameFromArg()
-	if err != nil {
-		return errors.Wrap(err, errors.New(err.Error()+"\nUse --client-only to query this client."))
-	}
 	return cmd.CheckEmpty(args[1:])
 }
 
 func (c *showCloudCommand) Info() *cmd.Info {
 	return jujucmd.Info(&cmd.Info{
 		Name:    "show-cloud",
 		Args:    "<cloud name>",
-		Purpose: "Shows detailed information on a cloud.",
+		Purpose: "Shows detailed information for a cloud.",
 		Doc:     showCloudDoc,
 	})
 }
 
 func (c *showCloudCommand) Run(ctxt *cmd.Context) error {
 	var (
-		cloud *CloudDetails
-		err   error
+		localCloud *CloudDetails
+		localErr   error
 	)
-	if c.ControllerName == "" {
-		cloud, err = c.getLocalCloud()
-	} else {
-		cloud, err = c.getControllerCloud()
+	if c.BothClientAndController || c.ClientOnly {
+		if localCloud, localErr = c.getLocalCloud(); c.ClientOnly && localErr != nil {
+			return localErr
+		}
 	}
-	if err != nil {
-		return err
+
+	var (
+		remoteCloud *CloudDetails
+		remoteErr   error
+	)
+	if c.BothClientAndController || c.ControllerOnly {
+		if c.ControllerName == "" {
+			// The user may have specified the controller via a --controller option.
+			// If not, let's see if there is a current controller that can be detected.
+			c.ControllerName, remoteErr = c.MaybePromptCurrentController(ctxt, fmt.Sprintf("show cloud %q from", c.CloudName))
+		}
+		if c.ControllerName == "" && remoteErr == nil {
+			remoteErr = errors.New("Not showing a cloud from a controller: no controller specified.")
+		}
+		if remoteErr == nil {
+			remoteCloud, remoteErr = c.getControllerCloud()
+		}
 	}
 
-	displayCloud := cloud
-	displayCloud.CloudType = displayCloudType(displayCloud.CloudType)
-	if err := c.out.Write(ctxt, displayCloud); err != nil {
+	var displayErr error
+	showRemoteConfig := c.includeConfig
+	if localCloud != nil && remoteCloud != nil {
+		// It's possible that a local cloud named A is different to
+		// a remote cloud named A. If their types are different and we
+		// need to display config, we'd need to list each cloud type config.
+		// If the clouds' types are the same, we only need to list
+		// config once after the local cloud information.
```

Comment:
Is that true? The types for client and controller cloud could both be "openstack" and config for each could be totally different. Or "maas" etc.
This also applies to public clouds like "aws" as I could have a local version with bespoke config which overrides the public version.

I may have missed it - it there a test for this scenario?

Reply:
It is true that for individual instances of a particular cloud type the configuration values are different. However, if you look at the code of this command we do not show the values but the metadata for the configuration based on the cloud type.

For example, for 'aws' which is of type 'ec2':
```
The available config options specific to ec2 clouds are:
vpc-id:
  type: string
  description: Use a specific AWS VPC ID (optional). When not specified, Juju requires
    a default VPC or EC2-Classic features to be available for the account/region.
vpc-id-force:
  type: bool
  description: Force Juju to use the AWS VPC ID specified with vpc-id, when it fails
    the minimum validation criteria. Not accepted without vpc-id
```

Reply:
great, that makes sense \o/